### ISSUE TITLE
A question about Spectral Discriminator

### ISSUE THREAD

**Issue by Kittywyk** [jdg900/UID-FDK#2]
Hi, thanks for your work.

I got one question about the implementation of the Spectral Discriminator. I found that the input is detached from the computational graph once entering the get_fft_feature function. Consequently, how can the generator be optimized since the output of the generator is detached?
```
def get_fft_feature(x):
    x_rgb = x.detach()
    epsilon = 1e-8

    x_gray = RGB2gray(x_rgb)
    fft = torch.rfft(x_gray,2,onesided=False)
    fft += epsilon
    magnitude_spectrum = torch.log((torch.sqrt(fft[:,:,:,0]**2 + fft[:,:,:,1]**2 + 1e-10))+1e-10)
    magnitude_spectrum = shift(magnitude_spectrum)
    magnitude_spectrum = magnitude_spectrum.cpu().numpy()

    out = []
    for i in range(magnitude_spectrum.shape[0]):
        out.append(torch.from_numpy(azimuthalAverage(magnitude_spectrum[i])).float().unsqueeze(0))
    out = torch.cat(out, dim=0)
    
    out = (out - torch.min(out, dim=1, keepdim=True)[0]) / (torch.max(out, dim=1, keepdim=True)[0] - torch.min(out, dim=1, keepdim=True)[0])
    out = Variable(out, requires_grad=True).to(x.device)
    
    return out


class Spectral_Discriminator(nn.Module):
    def __init__(self, height):
        super(Spectral_Discriminator, self).__init__()
        self.thresh = int(height / (2*math.sqrt(2)))
        self.linear = nn.Linear(self.thresh, 1)
    
    def forward(self, input: torch.Tensor):
        az_fft_feature = get_fft_feature(input)
        az_fft_feature[torch.isnan(az_fft_feature)] = 0
        
        return self.linear(az_fft_feature[:,-self.thresh:])
```

**Comment by jdg900** [comment]
Hi, thank you for your interest. Through the line "out = Variable(out, requires_grad=True).to(x.device)", requests_grad is true again, so that the generator can be optimized again.

**Comment by Kittywyk** [comment]
Thanks for your reply.
I'm a bit confused, because what I understand is that once the tensor is detached from the computational graph, a new tensor that requires grad is created. Though the line "out = Variable(out, requires_grad=True).to(x.device)" sets the gradient again, the tensor "out" is not in the previous computational graph.
Anyway, I just want to check with you. If I'm wrong, please correct me. I will do the experiment and visualize the computational graph later to confirm.

Here is a related post on the Pytorch forum. [https://discuss.pytorch.org/t/w-detach-and-then-requires-grad-again/36495]([https://discuss.pytorch.org/t/w-detach-and-then-requires-grad-again/36495)
